shapeshift/web
# feat: fox/eth full multi account support

## Description

- Overview
  - [x] Account selection
  - [x] Use selected account data
- [x] Deposit
  - [x] Approve
  - [x] Confirm
  - [x] Use selected account data in all steps
  - [ ] Poll (regression test against account 0 on prod) - seems borked, possibly because of unstable WS
- [x] Withdraw
  - [x] Approve
  - [x] Confirm 
  - [x] Use selected account data in all steps
  - [ ] Poll (regression test against account 0 on prod) - seems borked, possibly because of unstable WS
- [x] Claim
  - [x] Claim from selected account 
  - [x] Poll (regression test against account 0 on prod)
- [x] Deep-linking
  - [x] Open modal at current account from account asset page
  - [x] Display the balance of current account in account select page
 

## Notice




## Pull Request Type


## Issue (if applicable)

- relates to the closed https://github.com/shapeshift/web/issues/2609
- relates to the closed https://github.com/shapeshift/web/issues/2610
- relates to https://github.com/shapeshift/web/issues/2615
- relates to https://github.com/shapeshift/web/issues/2730

## Risk

Tested with MetaMask (which has only account 0 support) but this could bring regressions for the FOX/ETH modal for single accounts, test accordingly

## Testing (with KeepKey / Native)

- Enable the FOX/ETH LP / Farming flags as well as MultiAccount

- Add an Ethereum account and send some ETH and FOX to it
- Stake some FOX into FOX LP, reload the app (to avoid websocket shenanigans) and open Fox Farming from the DeFi earn section
- Select account 1 and click deposit
- Notice account #1 is pre-selected at Deposit step, with the right amount to deposit 
- Deposit some LP tokens into Fox Farming, sign and broadcast it - it should be deposited from account #1
- Disconnect and reconnect wallet, then reopen the Fox Farming modal and select withdraw with account #1 selected in overview
- Notice account #1 is pre-selected at Withdraw step, with the right amount available to withdraw
- Withdraw some LP tokens from Fox Farming - it should be withdrawn from account #1

- The aforementioned flow is working both for FOX LP / Farming

- When accessing the ETH/FOX Pool account asset page for an account (e.g 1), the farming/LP balances should be the ones from that account
- Clicking the farming/LP rows on account asset page should open the modal for the current account page

- Ensure the aforementioned flow still works with the MultiAccount flag off

### Engineering



- Refer to top-level testing steps
- Double-check the logic looks sound in deposit, withdraw, overview and claim for all steps

### Operations

- Refer to top-level testing steps



## Screenshots 

#### Farming Deposit

<img width="539" alt="image" src="https://user-images.githubusercontent.com/17035424/191581701-95de2cc5-4450-43c1-8600-60f4a5c7bb97.png">
<img width="531" alt="image" src="https://user-images.githubusercontent.com/17035424/191581746-67735549-f333-4c27-9483-5371e127849f.png">
<img width="525" alt="image" src="https://user-images.githubusercontent.com/17035424/191581789-f5b196a2-b429-4579-9732-20fdf1fbd4e5.png">

#### Farming Withdraw

<img width="525" alt="image" src="https://user-images.githubusercontent.com/17035424/191582043-b0337f83-f903-42f7-8065-29121135d746.png">
<img width="519" alt="image" src="https://user-images.githubusercontent.com/17035424/191582096-4abd5ea7-20ac-4e3f-9572-fd077abd68d9.png">

#### LP Deposit

<img width="537" alt="image" src="https://user-images.githubusercontent.com/17035424/191582252-955c6b95-9c31-470c-bc1e-23ad726478b9.png">
<img width="523" alt="image" src="https://user-images.githubusercontent.com/17035424/191582336-f74b6f79-ef5d-49a7-a6f4-fbdc36e43f37.png">

#### LP Withdraw

<img width="536" alt="image" src="https://user-images.githubusercontent.com/17035424/191583958-322d9378-ef47-4448-88bd-2af30a0a7cb2.png">
<img width="532" alt="image" src="https://user-images.githubusercontent.com/17035424/191584008-d65950fe-6da0-4474-bbcd-4ed86ab61155.png">
<img width="538" alt="image" src="https://user-images.githubusercontent.com/17035424/191591071-81ac6fd7-47d4-485c-b439-a2d86e642562.png">
